react-native-git-upgrade/cliEntry.js

```diff
@@ -0,0 +1,239 @@
+  const generatorArgs = ['react:app', context.appName].concat(context.cliArgs._);
+  return new Promise((resolve) => env.run(generatorArgs, {upgrade: true, force: true}, resolve));
+}
+
+async function run(cliVersion, cliArgs) {
+  const context = {
+    tmpDir: path.resolve(os.tmpdir(), 'react-native-git-upgrade'),
+    generatorDir: path.resolve(process.cwd(), 'node_modules', 'react-native', 'local-cli', 'generator'),
+    cliVersion,
+    cliArgs
+  };
+
+  try {
+    log.info('Read project descriptors');
+    const {rnPak, pak} = readProjectDescriptors();
+    context.appName = pak.name;
+    context.currentVersion = rnPak.version;
+    context.declaredVersion = pak.dependencies['react-native'];
+    context.declaredReactVersion = pak.dependencies.react;
+
+    log.info('Check declared version');
+    checkDeclaredVersion(context);
+
+    log.info('Check matching versions');
+    checkMatchingVersions(context);
+
+    log.info('Check React peer dependency');
+    checkReactPeerDependency(context);
+
+    log.info('Check Git installation');
+    checkGitAvailable();
+
+    log.info('Get react-native version from NPM registry');
+    const versionOutput = await exec(context, 'npm view react-native@' + (context.cliVersion || 'latest') + ' version');
+
+    log.info('Check new version');
+    context.newVersion = checkNewVersion(context, versionOutput);
+
+    log.info('Setup temporary working directory');
+    await setupWorkingDir(context);
+
+    log.info('Configure Git environment');
+    configureGitEnv(context);
+
+    log.info('Init Git repository');
+    await exec(context, 'git init');
+
+    log.info('Add all files to commit');
+    await exec(context, 'git add .');
+
+    log.info('Commit pristine sources');
+    await exec(context, 'git commit -m "Project snapshot"');
+
+    log.info ('Create a tag before updating sources');
+    await exec(context, 'git tag project-snapshot');
+    context.sourcesUpdated = true;
+
+    log.info('Generate old version template');
+    await generateTemplates(context);
+
+    log.info('Add updated files to commit');
+    await exec(context, 'git add .');
+
+    log.info('Commit old version template');
+    await exec(context, 'git commit -m "Old version" --allow-empty');
+
+    log.info('Install the new version');
+    await exec(context, 'npm install --save react-native@' + context.newVersion);
+
+    log.info('Generate new version template');
+    await generateTemplates(context);
+
+    log.info('Add updated files to commit');
+    await exec(context, 'git add .');
+
+    log.info('Commit new version template');
+    await exec(context, 'git commit -m "New version" --allow-empty');
+
+    log.info('Generate the patch between the 2 versions');
+    const diffOutput = await exec(context, 'git diff HEAD~1 HEAD');
+
+    log.info('Write the patch on filesystem');
+    context.patchPath = path.resolve(context.tmpDir, `upgrade_${context.currentVersion}_${context.newVersion}.patch`);
```

Comment:
We need it the `catch` blocks too. If we do so, we'll need a reference outside the try/catch: `let patchPath;`... Not so pretty, and `context` is already our "reference keeper object".